File: src/lib/editor/embeds.ts
import { StateField, type EditorState, type Extension } from '@codemirror/state';
import { Decoration, EditorView, WidgetType, type DecorationSet } from '@codemirror/view';
import { backend } from '$lib/ipc';
import { createEmbedImage, embedStyleAttr } from '$lib/embedImage';
import { indexStore } from '$lib/state/index.svelte';
import {
  classifyEmbedTarget,
  isImageAttachment,
  resolveEmbedNameIn,
  resolveEmbedPathIn,
  scanEmbedsUtf16,
} from '$lib/wasm/exports';
import { refreshBrokenLinks } from './broken-links';
import { openEmbedLightbox } from './embedLightbox';
import {
  embedSizeCss,
  embedWidgetKey,
  planEmbeds,
  type EmbedPlan,
} from './embedPlan';

// ---------------------------------------------------------------------------
// Embed rendering (slice: embedded-images, ADR-0010)
//
// `embedBlocks()` is Sunstone's replacement for atomic-editor's `imageBlocks()`,
// which is REMOVED from the extension list (`extensions.ts`) so nothing
// double-renders. It is one `StateField` emitting exactly one POINT
// `Decoration.widget` per Embed and NEVER a replace:
//
//   * hiding the raw `![alt](x.png)` source, and revealing it again under the
//     cursor, is entirely `inlinePreview`'s job — it already replaces the whole
//     `Image` node on inactive lines (`inline-preview.js:553-572`) and treats no
//     line as active when Sunstone's patched `alwaysRender` flag is set in
//     `read` mode. We layer over it; we never fight it and never patch it;
//   * a BLOCK widget must come from a StateField (CM6 forbids ViewPlugin-sourced
//     block decorations), and one field serves the inline case too.
//
// The placement/render decisions are all in the pure, unit-tested `embedPlan.ts`
// (the `mermaidBlocks.ts` / `mermaid.ts` split). This module is the CM/DOM shell
// only: the field, the widget, the `dimensionCache`, the click gestures and the
// `cm-embed-*` styling. The `<img>` itself and the two placeholders use the
// SHARED class names (`embed-image` / `embed-broken` / `embed-remote`) and the
// shared `createEmbedImage` builder, so the editor and the server render
// (`src/lib/rendered.css`, `src/lib/web/remoteEmbed.ts`) cannot drift.
//
// OFFSETS: the field scans with `scanEmbedsUtf16`, not `scanEmbeds`. The kernel
// reports BYTE offsets by default — the unit the SSR renderer and the rewrite
// engine slice Rust strings with — and CodeMirror positions are UTF-16 code
// units, so a byte offset puts every decoration after the first non-ASCII
// character in the document in the wrong place. The UTF-16 variant lives in
// `crates/sunstone-shared/src/embed.rs` beside the byte one, exactly as
// `critic.rs` and `citations.rs` already report this seam (ADR 0006 §4).
// ---------------------------------------------------------------------------

/**
 * Session-lifetime cache of observed NATURAL image dimensions, keyed by URL —
 * upstream `imageBlocks`' trick, kept verbatim and for its original reason.
 *
 * CM6's virtualizer unmounts line DOM when it leaves the viewport and calls
 * `toDOM` again on the way back. Without the cache an `<img>` remounts with no
 * intrinsic size, lays out as a zero-height box and snaps to its real size once
 * decode completes — the heightmap grows under an in-flight scroll, which on iOS
 * reads as an anchor conflict and halts kinetic scrolling. Setting the `width` /
 * `height` ATTRIBUTES from this cache pins the aspect ratio on mount.
 *
 * It stores natural dimensions ONLY. An explicit `|300` size is applied as CSS
 * (see `embedSizeCss`), so two differently-sized Embeds of one file cannot
 * corrupt each other's entry (ADR-0010).
 */
const dimensionCache = new Map<string, { w: number; h: number }>();

/** Record an image's natural size the first time we see it decoded. */
function rememberDimensions(img: HTMLImageElement, src: string): void {
  if (img.naturalWidth > 0 && img.naturalHeight > 0 && !dimensionCache.has(src)) {
    dimensionCache.set(src, { w: img.naturalWidth, h: img.naturalHeight });
  }
}

/**
 * The error placeholder, for BOTH failure states ADR-0010 collapses into one
 * visual: an unresolvable target, and a resolved target whose bytes fail to
 * load (the second is the one upstream has no handler for at all). Styled like
 * `cm-broken-link` — `var(--danger)`, dashed.
 *
 * The markup mirrors what the shared Rust renderer emits for the web view and
 * the print export — `<span class="embed-broken" data-broken="true"
 * data-embed-target=…>Missing attachment: …</span>`, styled in
 * `src/lib/rendered.css` — so the two surfaces read identically. The extra
 * `cm-embed-broken` is only this file's `EditorView.theme` hook, since
 * `rendered.css` scopes its rules under `.rendered`.
 */
function errorPlaceholder(alt: string, target: string): HTMLElement {
  const el = document.createElement('span');
  el.className = 'cm-embed-broken embed-broken';
  el.dataset.broken = 'true';
  el.dataset.embedTarget = target;
  el.textContent = `Missing attachment: ${alt}`;
  el.title = `No Attachment resolves ${target}`;
  return el;
}

/**
 * The neutral click-to-load affordance for a remote (`http`/`https`) Embed.
 * Opening a Concept must not fire a tracking pixel from a reader's browser, so
 * NOTHING is fetched here: the `<img>` is created, and its `src` set, only when
 * this button is clicked. Deliberately NOT an error visual — a remote Embed is a
 * normal state.
 */
function remoteAffordance(plan: EmbedPlan, load: () => void): HTMLElement {
  const button = document.createElement('button');
  button.type = 'button';
  button.className = 'cm-embed-remote embed-remote';
  // The same `data-embed-*` vocabulary the shared renderer puts on ITS button,
  // so one stylesheet and one mental model serve both surfaces — and, like it,
  // NO `src` anywhere until the reader asks.
  button.dataset.embedSrc = plan.src;
  button.dataset.embedAlt = plan.alt;
  const requested = embedSizeCss(plan.width, plan.height);
  const style = embedStyleAttr(requested.width, requested.height);
  if (style) button.dataset.embedStyle = style;
  button.textContent = `Load remote image — ${hostOf(plan.src)}`;
  button.title = plan.src;
  button.addEventListener('click', (event) => {
    event.preventDefault();
    // Keep the click off the wrapper's caret/lightbox handler.
    event.stopPropagation();
    load();
  });
  button.addEventListener('mousedown', (event) => event.stopPropagation());
  return button;
}

/** The host of a remote URL, for the affordance label. Falls back to the URL. */
function hostOf(url: string): string {
  try {
    return new URL(url).host || url;
  } catch {
    return url;
  }
}

/**
 * The point widget for one Embed. `eq()` is keyed on the pure
 * `embedWidgetKey` — `(render, placement, width, height, alt, src)` — so an
 * unrelated edit produces an EQUAL widget, CodeMirror reuses the element, and
 * the browser never re-decodes the image (ADR-0010; the caching lesson ADR-0005
 * records, in the form this field needs it).
 */
class EmbedWidget extends WidgetType {
  readonly key: string;

  constructor(
    readonly plan: EmbedPlan,
    readonly reading: boolean,
  ) {
    super();
    this.key = embedWidgetKey(plan);
  }

  eq(other: EmbedWidget): boolean {
    return other.key === this.key && other.reading === this.reading;
  }

  /**
   * The `<img>` for a resolved (or just-loaded remote) Embed, built by the
   * SHARED `createEmbedImage` — the same builder `src/lib/web/remoteEmbed.ts`
   * uses for the server-rendered surface, so the `embed-image` class, the lazy
   * loading, the "a requested size is CSS, never HTML attributes" rule and the
   * "assign `src` last" ordering cannot drift between the two.
   *
   * An explicit `|300` / `300x200` becomes CSS. A width alone leaves
   * `height: auto` from the theme, so the ratio holds; a width larger than the
   * natural width UPSCALES (Obsidian's behaviour) while `max-width: 100%` still
   * lets the content column clamp it.
   */
  private image(src: string, host: HTMLElement): HTMLImageElement {
    const requested = embedSizeCss(this.plan.width, this.plan.height);
    return createEmbedImage({
      src,
      alt: this.plan.alt,
      style: embedStyleAttr(requested.width, requested.height),
      natural: dimensionCache.get(src) ?? null,
      onLoad: (img) => rememberDimensions(img, src),
      // Upstream has no error handling at all: a resolved path whose bytes fail
      // to load would render as the browser's broken-image glyph. ADR-0010
      // folds that into the SAME placeholder an unresolvable target gets.
      onError: () => {
        host.replaceChildren(errorPlaceholder(this.plan.alt, this.plan.target));
        host.classList.add('cm-embed-failed');
      },
    });
  }

  toDOM(view: EditorView): HTMLElement {
    const block = this.plan.placement === 'block';
    const wrap = document.createElement(block ? 'div' : 'span');
    wrap.className = `cm-embed ${block ? 'cm-embed-block' : 'cm-embed-inline'}`;
    if (this.reading && this.plan.render === 'image') wrap.classList.add('cm-embed-zoomable');

    if (this.plan.render === 'error') {
      wrap.appendChild(errorPlaceholder(this.plan.alt, this.plan.target));
    } else if (this.plan.render === 'remote') {
      const src = this.plan.src;
      wrap.appendChild(
        remoteAffordance(this.plan, () => {
          wrap.replaceChildren(this.image(src, wrap));
        }),
      );
    } else {
      wrap.appendChild(this.image(this.plan.src, wrap));
    }

    // --- Click gesture -----------------------------------------------------
    // `editing`: upstream's caret-to-source. The source line is hidden by
    // `inlinePreview`, so this is the only way to reach it. The position is
    // resolved from the widget's CURRENT DOM location at event time
    // (`posAtDOM`), never from a captured offset, so it stays correct after an
    // unrelated edit shifted the range.
    //
    // `read`: there is no caret to place, so the gesture buys a lightbox — but
    // only for an image the content column actually DOWNSCALED. A picture shown
    // at or above its natural size has nothing more to reveal.
    if (this.reading) {
      wrap.addEventListener('click', (event) => {
        const img = wrap.querySelector('img');
        if (!(img instanceof HTMLImageElement) || img.naturalWidth === 0) return;
        if (img.clientWidth >= img.naturalWidth) return;
        event.preventDefault();
        event.stopPropagation();
        openEmbedLightbox(img.currentSrc || img.src, this.plan.alt);
      });
    } else {
      wrap.addEventListener('mousedown', (event) => {
        event.preventDefault();
        event.stopPropagation();
        const pos = view.posAtDOM(wrap);
        if (pos < 0) return;
        // A block widget sits at `line.to` with `side: 1`, so CM6 would map a
        // click to the START of the NEXT line; step back into the source line.
        // An inline widget already sits at the Embed's end, on the source line.
        const target = block ? Math.max(0, pos - 1) : pos;
        view.focus();
        view.dispatch({ selection: { anchor: target }, scrollIntoView: false });
      });
    }

    return wrap;
  }

  /** Our listeners above are the sole decider of what a pointer event means. */
  ignoreEvent(event: Event): boolean {
    return event.type === 'mousedown' || event.type === 'click';
  }
}

/**
 * Build the widget set for the current state: scan the body for Embeds (UTF-16
 * offsets), plan each one, and emit a block or inline point widget at the
 * planned anchor. `Decoration.set(_, true)` sorts, since block widgets anchor at
 * `line.to` while inline ones anchor at the Embed's end.
 */
function buildEmbeds(
  state: EditorState,
  reading: boolean,
  currentPath: () => string,
): DecorationSet {
  const embeds = scanEmbedsUtf16(state.doc.toString());
  if (embeds.length === 0) return Decoration.none;

  const source = currentPath();
  // The Attachment corpus is a SEPARATE index from the Concept path set (an
  // Attachment is never a Concept), read fresh per build so a newly added image
  // resolves on the next rebuild.
  const attachments = indexStore.attachmentPaths();

  const plans = planEmbeds(embeds, state.doc, {
    reading,
    selection: state.selection.ranges,
    classify: classifyEmbedTarget,
    isImage: isImageAttachment,
    // The name model searches the Attachment corpus, so a miss is known here
    // and becomes the error placeholder immediately. The path model is pure
    // path math and deliberately does NOT consult `indexStore.attachmentExists`:
    // the Attachment index is empty until the first refresh, and gating on it
    // would flash every local Embed as broken on the first frame. A path with
    // nothing behind it still yields a URL, the `<img>` fails, and `onError`
    // shows the SAME placeholder — which is exactly what ADR-0010 asks for.
    resolve: (embed) =>
      embed.kind === 'name'
        ? resolveEmbedNameIn(attachments, embed.target)
        : resolveEmbedPathIn(source, embed.target),
    // Synchronous on purpose — a decoration builder cannot await (see the
    // contract on `Backend.attachmentUrl`).
    url: (path) => backend.attachmentUrl(path),
  });

  return Decoration.set(
    plans.map((plan) =>
      Decoration.widget({
        widget: new EmbedWidget(plan, reading),
        block: plan.placement === 'block',
        // `side: 1` puts a block widget AFTER its line's content (the image sits
        // below its source) and an inline widget after the Embed's last
        // character (so it follows the text it belongs to).
        side: 1,
      }).range(plan.at),
    ),
    true,
  );
}

/**
 * The Embed StateField. Rebuilds on:
 *   - a doc change (an Embed may have been added, removed, resized or retargeted);
 *   - a selection change (the inline cursor-overlap reveal);
 *   - `refreshBrokenLinks`, which the host already dispatches on the
 *     `file-changed` watcher event and on a Concept switch — exactly the events
 *     that can change whether an Attachment resolves.
 */
function embedField(reading: boolean, currentPath: () => string): StateField<DecorationSet> {
  return StateField.define<DecorationSet>({
    create: (state) => buildEmbeds(state, reading, currentPath),
    update(deco, tr) {
      for (const effect of tr.effects) {
        if (effect.is(refreshBrokenLinks)) {
          return buildEmbeds(tr.state, reading, currentPath);
        }
      }
      if (tr.docChanged || tr.selection) {
        return buildEmbeds(tr.state, reading, currentPath);
      }
      return deco;
    },
    provide: (f) => EditorView.decorations.from(f),
  });
}

/**
 * Embed styling.
 *
 * The visuals are deliberately the SAME two `src/lib/rendered.css` gives the
 * server-rendered surface — the broken placeholder mirrors `cm-broken-link`
 * (`var(--danger)`, dashed), the remote affordance is neutral because a remote
 * Embed is a normal state and not an error — but the rules have to be restated
 * here: `rendered.css` scopes everything under `.rendered`, which the editor's
 * DOM is not. Keep the two in step; the class names are the contract.
 */
export const embedTheme = EditorView.theme({
  '.cm-embed-block': {
    display: 'block',
    padding: '0.25rem 0',
  },
  '.cm-embed-inline': {
    display: 'inline-block',
    verticalAlign: 'text-bottom',
  },
  // `max-width: 100%` keeps an explicit size a REQUEST the content column may
  // clamp; `height: auto` preserves the ratio when only a width was asked for.
  // An inline explicit `style.width` overrides both.
  '.cm-embed .embed-image': {
    maxWidth: '100%',
    height: 'auto',
    borderRadius: 'var(--radius-sm, 4px)',
  },
  // Reading mode: a click may open the lightbox, so hint at it.
  '.cm-embed-zoomable .embed-image': {
    cursor: 'zoom-in',
  },
  '.cm-embed-broken': {
    display: 'inline-block',
    padding: '0.05em 0.4em',
    border: '1px dashed var(--danger, #c0392b)',
    borderRadius: 'var(--radius-sm, 4px)',
    background: 'var(--danger-soft, rgba(192, 57, 43, 0.08))',
    color: 'var(--danger, #c0392b)',
    fontSize: '0.85em',
    cursor: 'help',
  },
  '.cm-embed-remote': {
    display: 'inline-flex',
    alignItems: 'center',
    maxWidth: '100%',
    padding: '0.2em 0.6em',
    border: '1px dashed var(--border, #bbb)',
    borderRadius: 'var(--radius-sm, 4px)',
    background: 'var(--bg-sunken, rgba(127, 127, 127, 0.08))',
    color: 'var(--text-muted, #777)',
    font: 'inherit',
    fontSize: '0.85em',
    lineHeight: '1.6',
    cursor: 'pointer',
    overflow: 'hidden',
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis',
  },
  '.cm-embed-remote:hover': {
    background: 'var(--hover, rgba(127, 127, 127, 0.15))',
    color: 'var(--text, inherit)',
  },
});

/**
 * The Embed extension (ADR-0010). Wire into `modeExtensions` for BOTH modes,
 * BEFORE `inlinePreview` — it is that extension which hides the raw source this
 * field renders over, and which we must not overlap.
 *
 * @param reading     `read` mode: the inline cursor-overlap reveal is off (there
 *                    is no caret) and a click opens the lightbox instead of
 *                    placing the caret.
 * @param currentPath the open Concept's bundle-relative path, read per build —
 *                    the base a path-model Embed (`![alt](./x.png)`) resolves
 *                    against.
 */
export function embedBlocks(reading: boolean, currentPath: () => string): Extension {
  return [embedField(reading, currentPath), embedTheme];
}

File: crates/sunstone-native/src/render/embeds.rs
//! The **Embed** pass of the SSR render (ei-1, ADR-0010/ADR-0011).
//!
//! An Embed (`![alt](path.png)` / `![[name.png]]`) points at an **Attachment** —
//! a non-`.md` file in the Bundle. comrak emits the author's raw relative `src`
//! for the first form and does not know the second at all, so both 404 in the
//! web viewer and in the desktop print/PDF window. This pass gives every Embed a
//! `src` the surrounding shell can actually fetch.
//!
//! ## Why a mapper and not a prefix
//!
//! ADR-0011 wrote "an asset-URL **prefix** parameter". That turned out to be
//! wrong: the two shells' URL shapes are not prefix-compatible.
//!
//! * desktop — `sunstone-asset://localhost/<encodeURIComponent(path)>`, where
//!   the WHOLE path is one percent-encoded segment (`a/b.png` → `a%2Fb.png`);
//! * web — `/api/asset?path=<percent-encoded path>`, a query value.
//!
//! No string prefix produces both, so the renderer takes a
//! `&dyn Fn(&str) -> String` mapping a bundle-relative path to a URL and each
//! shell supplies its own. See [`super::render_concept`]'s callers.
//!
//! ## Two stages, mirroring `critic.rs` / `citations.rs`
//!
//! `render.unsafe_` is deliberately `false` in this pipeline, so nothing may be
//! injected as raw HTML *into the markdown*. Instead:
//!
//! 1. **Before comrak** ([`embeds_to_markers`]) every Embed found by
//!    [`embed::scan_embeds`] is replaced — in the markdown stream, by byte span —
//!    with `![](sapembed:N)`, an ordinary markdown image whose destination is an
//!    index into a side table. This is also how `![[ … ]]` reaches comrak at all:
//!    `wikilink::replace_wikilinks` deliberately SKIPS embeds (see the asymmetry
//!    note in `sunstone_shared::embed`) and `wikilink.rs` stays untouched.
//! 2. **After comrak** ([`rewrite_embed_markers`]) each emitted
//!    `<img src="sapembed:N" …>` is replaced by the final markup for table entry
//!    `N` — exactly the technique `rewrite_marker_hrefs` uses for links.
//!
//! Every value interpolated into that markup is attribute-escaped here, since it
//! is author-controlled text arriving after comrak has stopped escaping.
//!
//! ## What renders as what (ADR-0011's failure states)
//!
//! | target | result |
//! |---|---|
//! | in-Bundle image that exists | `<img class="embed-image" src="<mapped>">` |
//! | in-Bundle image that does not | `<span class="embed-broken" data-broken="true">` |
//! | `http:` / `https:` | `<button class="embed-remote">` — click-to-load, never auto-fetched |
//! | `data:` | nothing at all |
//! | any other `scheme:`, or a non-image extension | left exactly as the author wrote it (out of scope until `al-1`) |
//!
//! The remote case is the one with teeth: on the web shell the author and the
//! reader are different people, so opening a Concept must not fire a tracking
//! pixel from the reader's browser and IP.

use std::sync::LazyLock;

use regex::Regex;

use sunstone_shared::embed::{self, EmbedKind, EmbedTargetKind};

/// The marker scheme an Embed's destination carries through comrak. Chosen to
/// match the `sap…:` family the link markers already use.
const M_EMBED: &str = "sapembed:";

/// What one Embed renders as, once resolved. One entry per marker index.
#[derive(Debug, Clone, PartialEq, Eq)]
pub(super) enum EmbedRender {
    /// A resolved in-Bundle Attachment: a real `<img>` at a shell-mapped URL.
    Image {
        /// The URL the shell's mapper produced.
        src: String,
        /// The accessible name the kernel already resolved (`Embed.alt`).
        alt: String,
        /// `width` / `height` as a CSS declaration list, never HTML attributes —
        /// an HTML attribute would corrupt the editor's URL-keyed
        /// `dimensionCache`, which stores NATURAL dimensions (ticket decision).
        style: Option<String>,
    },
    /// Unresolvable: a visible placeholder, never an abort.
    Broken {
        /// What the author asked for, shown so the mistake is findable.
        target: String,
        alt: String,
    },
    /// `http:` / `https:`: a neutral click-to-load affordance that fetches
    /// nothing until a reader asks for it.
    Remote {
        src: String,
        alt: String,
        style: Option<String>,
    },
    /// A `data:` URI — renders nothing at all.
    Dropped,
}

/// Build the CSS declarations for a requested size, or `None` when the author
/// asked for no size. Height without width is not expressible in the Embed
/// syntax (`|300x200` always carries the width), so only these two shapes exist.
fn size_style(width: Option<u32>, height: Option<u32>) -> Option<String> {
    match (width, height) {
        (Some(w), Some(h)) => Some(format!("width:{w}px;height:{h}px")),
        (Some(w), None) => Some(format!("width:{w}px")),
        _ => None,
    }
}

/// Replace every Embed in `body` with a `![](sapembed:N)` marker image, and
/// return the rewritten body plus the side table the markers index.
///
/// * `source_path` — the embedding Concept, for relative path resolution;
/// * `attachments` — every Attachment path in the Bundle (the separate index
///   ei-1 keeps apart from `all_paths`), used BOTH to resolve `![[name.png]]` by
///   name and to decide whether a path-resolved Embed actually exists;
/// * `asset_url` — the shell's bundle-relative-path → URL mapper.
///
/// Runs on the raw (frontmatter-stripped) body, BEFORE the CriticMarkup and
/// citation sentinel passes: those rewrite bytes, and an Embed's alt text must
/// not pick up a private-use sentinel that a later pass would expand into HTML
/// tags inside an attribute value.
pub(super) fn embeds_to_markers(
    body: &str,
    source_path: &str,
    attachments: &[String],
    asset_url: &dyn Fn(&str) -> String,
) -> (String, Vec<EmbedRender>) {
    let found = embed::scan_embeds(body);
    if found.is_empty() {
        return (body.to_string(), Vec::new());
    }

    let mut table: Vec<EmbedRender> = Vec::new();
    let mut out = String::with_capacity(body.len());
    let mut cursor = 0usize;

    for e in found {
        let render = match embed::classify_target(&e.target) {
            EmbedTargetKind::Data => Some(EmbedRender::Dropped),
            EmbedTargetKind::Remote => Some(EmbedRender::Remote {
                src: e.target.trim().to_string(),
                alt: e.alt.clone(),
                style: size_style(e.width, e.height),
            }),
            // `mailto:`, `file:`, `c:/…` — not an Attachment and not something
            // this ticket renders. Left byte-for-byte as the author wrote it.
            EmbedTargetKind::OtherScheme => None,
            EmbedTargetKind::Local => {
                if !embed::is_image_path(&e.target) {
                    // A non-image Attachment (and a `![[concept]]` transclusion)
                    // keeps today's literal rendering until `al-1`.
                    None
                } else {
                    let resolved = match e.kind {
                        EmbedKind::Path => embed::resolve_embed_path(source_path, &e.target),
                        EmbedKind::Name => embed::resolve_embed_name(attachments, &e.target),
                    };
                    Some(match resolved {
                        Some(path) if attachments.iter().any(|a| a == &path) => EmbedRender::Image {
                            src: asset_url(&path),
                            alt: e.alt.clone(),
                            style: size_style(e.width, e.height),
                        },
                        // Resolved but absent, or not resolvable at all: the
                        // author still sees WHAT was asked for.
                        Some(path) => EmbedRender::Broken {
                            target: path,
                            alt: e.alt.clone(),
                        },
                        None => EmbedRender::Broken {
                            target: e.target.trim().to_string(),
                            alt: e.alt.clone(),
                        },
                    })
                }
            }
        };

        // An author who literally writes `![](sapembed:0)` would otherwise hand
        // themselves another Embed's markup (a display quirk, not an escape —
        // every table entry comes from the same document). Drop it instead.
        let render = if e.target.trim_start().starts_with(M_EMBED) {
            Some(EmbedRender::Dropped)
        } else {
            render
        };

        let Some(render) = render else { continue };
        out.push_str(&body[cursor..e.from]);
        out.push_str(&format!("![]({M_EMBED}{})", table.len()));
        table.push(render);
        cursor = e.to;
    }
    out.push_str(&body[cursor..]);

    (out, table)
}

/// Replace each `<img src="sapembed:N" …>` comrak emitted with the final markup
/// for table entry `N`.
///
/// Everything emitted here is INLINE-level (`<span>` / `<button>` / `<img>`):
/// comrak puts a lone image inside a `<p>`, and a block element there would make
/// the browser close the paragraph early and reflow the rest of the Concept.
pub(super) fn rewrite_embed_markers(html: &str, table: &[EmbedRender]) -> String {
    if table.is_empty() {
        return html.to_string();
    }
    static RE: LazyLock<Regex> =
        LazyLock::new(|| Regex::new(r#"<img src="sapembed:(\d+)"[^>]*>"#).unwrap());
    RE.replace_all(html, |caps: &regex::Captures| {
        let idx: usize = caps[1].parse().unwrap_or(usize::MAX);
        match table.get(idx) {
            Some(r) => render_markup(r),
            // Unreachable for markers this module produced; leaving the tag in
            // place is a visible-but-harmless failure rather than a panic.
            None => caps[0].to_string(),
        }
    })
    .into_owned()
}

/// The final HTML for one table entry. Every interpolated value is
/// attribute-escaped: this runs after comrak has stopped escaping, and the
/// values are author-controlled.
fn render_markup(r: &EmbedRender) -> String {
    match r {
        EmbedRender::Image { src, alt, style } => format!(
            r#"<img class="embed-image" src="{}" alt="{}" loading="lazy"{} />"#,
            attr(src),
            attr(alt),
            style_attr(style),
        ),
        EmbedRender::Broken { target, alt } => format!(
            r#"<span class="embed-broken" data-broken="true" data-embed-target="{}" title="Attachment not found: {}">Missing attachment: {}</span>"#,
            attr(target),
            attr(target),
            attr(alt),
        ),
        // The URL rides in `data-embed-src`, NOT in `src`: the whole point is
        // that the browser must not fetch it until a reader clicks.
        EmbedRender::Remote { src, alt, style } => format!(
            r#"<button type="button" class="embed-remote" data-embed-src="{}" data-embed-alt="{}" data-embed-style="{}" title="{}">Load remote image</button>"#,
            attr(src),
            attr(alt),
            attr(style.as_deref().unwrap_or("")),
            attr(src),
        ),
        EmbedRender::Dropped => String::new(),
    }
}

/// ` style="…"`, or nothing.
fn style_attr(style: &Option<String>) -> String {
    match style {
        Some(s) => format!(r#" style="{}""#, attr(s)),
        None => String::new(),
    }
}

/// Escape for an HTML double-quoted attribute value (and for text content, whose
/// needs are a subset). A thin alias so this module reads as self-contained.
fn attr(s: &str) -> String {
    super::attr_escape(s)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn atts(ps: &[&str]) -> Vec<String> {
        ps.iter().map(|s| s.to_string()).collect()
    }

    fn web(path: &str) -> String {
        format!(
            "/api/asset?path={}",
            sunstone_shared::url::query_encode(path)
        )
    }

    fn markers(body: &str, source: &str, attachments: &[&str]) -> (String, Vec<EmbedRender>) {
        embeds_to_markers(body, source, &atts(attachments), &web)
    }

    #[test]
    fn a_markdown_embed_becomes_a_marker_and_a_table_entry() {
        let (out, table) = markers("![a](img.png)", "a.md", &["img.png"]);
        assert_eq!(out, "![](sapembed:0)");
        assert_eq!(
            table,
            vec![EmbedRender::Image {
                src: "/api/asset?path=img.png".into(),
                alt: "a".into(),
                style: None,
            }]
        );
    }

    #[test]
    fn a_wikilink_embed_resolves_by_name() {
        let (out, table) = markers("x ![[logo.png]] y", "deep/a.md", &["assets/logo.png"]);
        assert_eq!(out, "x ![](sapembed:0) y");
        assert_eq!(
            table,
            vec![EmbedRender::Image {
                src: web("assets/logo.png"),
                alt: "logo.png".into(),
                style: None,
            }]
        );
    }

    #[test]
    fn a_non_image_embed_is_left_untouched() {
        // Out of scope until `al-1`: a non-image Attachment and a `![[concept]]`
        // transclusion both keep today's literal rendering.
        let (out, table) = markers("![d](notes.pdf) ![[other]]", "a.md", &["notes.pdf"]);
        assert_eq!(out, "![d](notes.pdf) ![[other]]");
        assert!(table.is_empty());
    }

    #[test]
    fn an_other_scheme_embed_is_left_untouched() {
        let (out, table) = markers("![m](mailto:a@b.c)", "a.md", &[]);
        assert_eq!(out, "![m](mailto:a@b.c)");
        assert!(table.is_empty());
    }

    #[test]
    fn a_forged_marker_target_is_dropped_rather_than_impersonating_an_entry() {
        let (_, table) = markers("![](sapembed:0) ![x](img.png)", "a.md", &["img.png"]);
        assert_eq!(table[0], EmbedRender::Dropped);
    }

    #[test]
    fn markup_never_emits_a_block_element() {
        // comrak wraps a lone image in a `<p>`; a block element there would close
        // the paragraph early in the browser.
        for r in [
            EmbedRender::Image {
                src: "u".into(),
                alt: "a".into(),
                style: None,
            },
            EmbedRender::Broken {
                target: "t".into(),
                alt: "a".into(),
            },
            EmbedRender::Remote {
                src: "https://e/x.png".into(),
                alt: "a".into(),
                style: None,
            },
        ] {
            let html = render_markup(&r);
            for block in ["<div", "<p", "<figure", "<section"] {
                assert!(!html.contains(block), "{html} contains {block}");
            }
        }
    }

    #[test]
    fn author_text_is_attribute_escaped_in_the_emitted_markup() {
        let html = render_markup(&EmbedRender::Broken {
            target: r#"a"><script>x</script>"#.into(),
            alt: "alt".into(),
        });
        assert!(!html.contains("<script>"));
        assert!(html.contains("&quot;&gt;&lt;script&gt;"));
    }
}
